cmd/manager/cleanup.go

```diff
@@ -0,0 +1,178 @@
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// NewCleanUpResources returns a CleanUpCrew that updates existing
+// subscriptions linked to old CSCs and links them to the datastore CSCs
+// in the openshift-marketplace namespace, and then deletes those CSCs
+// along with existing datastore CSCs.
+func NewCleanUpResources(logger *logrus.Entry, client client.Client) CleanUpCrew {
+	return CleanUpCrew{
+		logger: logger,
+		client: client,
+	}
+}
+
+type CleanUpCrew struct {
+	logger *logrus.Entry
+	client client.Client
+}
+
+func (c *CleanUpCrew) CleanupOldResource() error {
+
+	c.logger.Printf("Updating existing subscriptions.")
+	datastoreCscNames, err := updateSubsciptions(c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting datastore CatalogSourceConfigs.")
+	err = deleteDatastoreCscs(datastoreCscNames, c.client)
+	if err != nil {
+		return err
+	}
+
+	c.logger.Printf("Deleting old CatalogSourceConfigs created duing operator installations.")
+	err = deleteOldCscs(c.client)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func updateSubsciptions(kubeClient client.Client) ([]string, error) {
+	var cscNames []string
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("csc-owner-name"))
+
+	subscriptions := &olm.SubscriptionList{}
+	// Get the list of existing subscriptions that have the label "csc-owner-name"
+	err := kubeClient.List(context.TODO(), options, subscriptions)
+	if err != nil {
+		return nil, err
+	}
+	for _, instance := range subscriptions.Items {
+		cscNames = append(cscNames, instance.GetLabels()["csc-owner-name"])
+		// try to guess the datastore catalogsource from the subscription
+		datastoreCs, err := findCatalogSource(&instance, kubeClient)
+		if k8s_errors.IsNotFound(err) {
+			// guess the catalogsource from the opsrc that has the package
+			datastoreCs, err = findCsFromOpsrc(&instance, kubeClient)
+			if err != nil {
+				return nil, err
+			}
+		}
+		// point the subscription to the datastore CatalogSource
+		instance.SetCatalogSource(datastoreCs.GetName())
+		instance.SetCatalogSourceNamespace("openshift-marketplace")
+		kubeClient.Update(context.TODO(), &instance)
+	}
+	return cscNames, nil
+}
+
+func deleteDatastoreCscs(cscNames []string, kubeClient client.Client) error {
+	for _, cscName := range cscNames {
+		namespacedName := types.NamespacedName{Name: cscName, Namespace: "openshift-marketplace"}
+		csc := &marketplace.CatalogSourceConfig{}
+		err := kubeClient.Get(context.TODO(), namespacedName, csc)
+		if err == nil {
+			err = kubeClient.Delete(context.TODO(), csc)
+			if err != nil {
+				return err
+			}
+		} else if err != nil && !k8s_errors.IsNotFound(err) {
+			return err
+		}
+
+	}
+	return nil
+}
+
+func deleteOldCscs(kubeClient client.Client) error {
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("opsrc-datastore: \"true\""))
```

Comment:
This seems to imply that you are looking for and deleting the datastore `CatalogSourceConfigs` here but you call this function `deleteOldCscs`. This is very confusing.